File: app/src/main/java/com/testtarget/ui/repo/TopRepoListPresenter.java
package com.testtarget.ui.repo;

import android.support.annotation.VisibleForTesting;
import android.view.View;

import com.testtarget.network.model.TopWeekly;
import com.testtarget.network.repository.Repository;
import com.testtarget.utils.schedulers.SchedulerProvider;

import java.util.List;

/**
 * Created by kuljeetsingh on 12/6/18.
 */

public class TopRepoListPresenter {
    private Repository repository;
    protected TopRepoListViewModel viewModel;
    private final String LANGUAGE = "java";
    private final String SINCE_TYPE = "weekly";
    private  SchedulerProvider schedulerProvider;

    public TopRepoListPresenter(TopRepoListViewModel viewModel, Repository repository, SchedulerProvider schedulerProvider) {
        this.repository = repository;
        this.viewModel = viewModel;
        this.schedulerProvider = schedulerProvider;
        setupListErrorVisibility(View.GONE, View.GONE, View.VISIBLE);
    }

    @VisibleForTesting
    SchedulerProvider getSchedulerProvider() {
        return schedulerProvider;
    }



    private void setupListErrorVisibility(int errorVisibility, int listVisibility, int loadingVisibility) {

        viewModel.errorMessageVisibility.setValue(errorVisibility);
        viewModel.repoListVisibility.setValue(listVisibility);
        viewModel.loadingVisibility.setValue(loadingVisibility);
    }



    public void getTopWeeklyRepo() {
        repository.getTopWeeklyRepo(LANGUAGE, SINCE_TYPE)
                .subscribeOn(getSchedulerProvider().io())
                .observeOn(getSchedulerProvider().ui())
                .subscribe(this::onRepoListSuccess, this::onNetworkFailure);
    }

    private void onNetworkFailure(Throwable throwable) {
        //you can handle network exceptions here like retry dialog show error messages
        setupListErrorVisibility(View.VISIBLE, View.GONE, View.GONE);
        viewModel.errorMessage.setValue("Network error. Please try again later.");
    }

    private void onRepoListSuccess(List<TopWeekly> topWeeklies) {
        if (topWeeklies != null && topWeeklies.size() > 0) {
            setupListErrorVisibility(View.GONE, View.VISIBLE, View.GONE);
            viewModel.repoListData.postValue(topWeeklies);
            return;
        }
        setupListErrorVisibility(View.VISIBLE, View.GONE, View.GONE);
        viewModel.errorMessage.setValue("No list available.");
    }
}

File: app/src/main/java/com/testtarget/ui/repo/TopRepoListPresenter.kt
package com.testtarget.ui.repo

import android.view.View
import com.testtarget.network.model.TopWeekly
import com.testtarget.network.repository.Repository
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

/**
 * Created by kuljeetsingh on 12/6/18.
 */

class TopRepoListPresenter(protected var viewModel: TopRepoListViewModel, private val repository: Repository) {
    private val LANGUAGE = "java"
    private val SINCE_TYPE = "weekly"

    init {
        setupListErrorVisibility(View.GONE, View.GONE, View.VISIBLE)
    }

    private fun setupListErrorVisibility(errorVisibility: Int, listVisibility: Int, loadingVisibility: Int) {
        viewModel.errorMessageVisibility.value = errorVisibility
        viewModel.repoListVisibility.value = listVisibility
        viewModel.loadingVisibility.value = loadingVisibility
    }

    fun getTopWeeklyRepo() {
        repository.getTopWeeklyRepo(LANGUAGE, SINCE_TYPE)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ this.onRepoListSuccess(it) }, { this.onNetworkFailure(it) })
    }

    private fun onNetworkFailure(throwable: Throwable) {
        //you can handle network exceptions here like retry dialog show error messages
        setupListErrorVisibility(View.VISIBLE, View.GONE, View.GONE)
        viewModel.errorMessage.value = "Network error. Please try again later."
    }

    private fun onRepoListSuccess(topWeeklies: List<TopWeekly>?) {
        if (topWeeklies != null && topWeeklies.isNotEmpty()) {
            setupListErrorVisibility(View.GONE, View.VISIBLE, View.GONE)
            viewModel.repoListData.postValue(topWeeklies)
            return
        }
        setupListErrorVisibility(View.VISIBLE, View.GONE, View.GONE)
        viewModel.errorMessage.value = "No list available."
    }
}
